{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'windows'"
      ],
      "text/latex": [
       "'windows'"
      ],
      "text/markdown": [
       "'windows'"
      ],
      "text/plain": [
       "[1] \"windows\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "'R version 3.5.1 (2018-07-02)'"
      ],
      "text/latex": [
       "'R version 3.5.1 (2018-07-02)'"
      ],
      "text/markdown": [
       "'R version 3.5.1 (2018-07-02)'"
      ],
      "text/plain": [
       "[1] \"R version 3.5.1 (2018-07-02)\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "'Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz'"
      ],
      "text/latex": [
       "'Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz'"
      ],
      "text/markdown": [
       "'Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz'"
      ],
      "text/plain": [
       "[1] \"Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "8"
      ],
      "text/latex": [
       "8"
      ],
      "text/markdown": [
       "8"
      ],
      "text/plain": [
       "[1] 8"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "8.59 GB"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "library(benchmarkme)\n",
    "get_platform_info()$OS.type\n",
    "get_r_version()$version.string\n",
    "get_cpu()$model_name;get_cpu()$no_of_cores\n",
    "get_ram()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "##Package was modified to allow complete cross mating\n",
    "#library(\"devtools\")\n",
    "#install_github(\"gglinzijie/xbreed\")\n",
    "library(\"xbreed\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Genome specification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "3928"
      ],
      "text/latex": [
       "3928"
      ],
      "text/markdown": [
       "3928"
      ],
      "text/plain": [
       "[1] 3928"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Number of markers per chr from chapter 1\n",
    "lin_map<-read.table(\"raw.map\")\n",
    "m=1:22\n",
    "for(i in 1:22) {m[i]=dim(lin_map[lin_map$V1==i,])[1]}\n",
    "sum(m)\n",
    "#nf <- layout(1,widths=3, heights=4, TRUE)\n",
    "#barplot(m,col = \"cadetblue1\",ylim=c(0,400),xlim=c(0,23),cex.axis = 1.5,names.arg = )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "#data from chaptre 1\n",
    "geno<-read.table(\"Hetero_realigned_cov10_filtered3.raw\")\n",
    "pheno<-read.csv(\"2017heteroPheno.csv\", header=T)\n",
    "attach(pheno)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#parameter of genome\n",
    "no.chr<-22\n",
    "genome<-data.frame(matrix(NA, nrow=no.chr, ncol=6))\n",
    "names(genome)<-c(\"chr\",\"len\",\"nmrk\",\"mpos\",\"nqtl\",\"qpos\")\n",
    "genome$chr<-c(1:no.chr) #Chromosome id from 1 to 22\n",
    "genome$len<-c(200,rep(100,21))#Chromosome length in cM\n",
    "genome$nmrk<-c(m) #Number of markers, 3928 in total \n",
    "genome$mpos<-c('even') \n",
    "genome$nqtl<-c(40) #Number of qtl  40*22 = 880 in total\n",
    "genome$qpos<-c('rnd')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Historiacal population"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "---sel_seq_qtl is missing, it has been set to default value of 0\n",
      "---sel_seq_mrk is missing, it has been set to default value of 0\n",
      "Historical pop is initialized...\n",
      "Simulating trait ...\n",
      "Output data preparation ...\n",
      "Establishment of historical population completed\n"
     ]
    }
   ],
   "source": [
    "#Historical population \n",
    "hp<-make_hp(hpsize=1000 ,ng=1000,h2=0.654,d2=0,phen_var=84,\n",
    "            genome=genome,mutr=2.5*10**-4,laf=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "0.333333333333333"
      ],
      "text/latex": [
       "0.333333333333333"
      ],
      "text/markdown": [
       "0.333333333333333"
      ],
      "text/plain": [
       "[1] 0.3333333"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "0.285962320819597"
      ],
      "text/latex": [
       "0.285962320819597"
      ],
      "text/markdown": [
       "0.285962320819597"
      ],
      "text/plain": [
       "[1] 0.2859623"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#validation\n",
    "mutr<-2.5*10**-4\n",
    "ne<-1000\n",
    "k<-2 \n",
    "Fneu<-4*ne*mutr\n",
    "(Expected_het1<-1-((1+((Fneu)/(k-1)))/(1+((Fneu*k)/(k-1)))))\n",
    "(het_observed<-mean(2*(hp$freqMrk[,3]*hp$freqMrk[,4])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "Male_founders<-data.frame(number=10,select='rnd')\n",
    "Female_founders<-data.frame(number=10,select='rnd')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "Selection<-data.frame(matrix(NA, nrow=2, ncol=3))\n",
    "names(Selection)<-c('Number','type','Value') \n",
    "Selection$Number[1:2]<-c(10,10)\n",
    "Selection$type[1:2]<-c('rnd','rnd')\n",
    "Selection$Value[1:2]<-c('l','l')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "sh_output<-data.frame(matrix(NA, nrow=1, ncol=4))\n",
    "names(sh_output)<-c(\"data\",\"qtl\",\"freq_mrk\",\"marker\")\n",
    "sh_output[1]<-c(1) \n",
    "sh_output[2]<-c(1) \n",
    "sh_output[3]<-c(1)\n",
    "sh_output[4]<-c(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Controlling input data ...\n",
      "Intializing base population ...\n",
      "Generation 0 started ......... \n",
      "Generation 0 is finished. Time taken: 17.03785\n",
      "Generation 1 started ......... \n",
      "Generation 1 is finished. Time taken: 15.43671\n",
      "Output data preparation ...\n",
      "Writing output files ...\n",
      "Sampling hp is done!\n"
     ]
    }
   ],
   "source": [
    "#10 sires and 10 dams perform complete cross; 200 eggs were produced per dam.\n",
    "RP<-sample_hp(hp_out=hp,Male_founders= Male_founders,\n",
    "              Female_founders=Female_founders,ng=1,Selection=Selection, \n",
    "              litter_size=200,saveAt=\"SNP3928\",sh_output=sh_output,Display=FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "#function for calculate the allele coding, which is 0, 1, 2\n",
    "bin_snp<-function(mat){\n",
    "s1<-seq(1,ncol(mat),2)\n",
    "s2<-seq(2,ncol(mat),2)\n",
    "a1<-mat[,s1]+mat[,s2]\n",
    "a1[a1==3]=1\n",
    "a1[a1==4]=0\n",
    "snp_code<-a1\n",
    "return(snp_code)\n",
    " }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "##geno and pheno from simulated data\n",
    "pheno<-RP$output[[2]]$data$phen\n",
    "n<-bin_snp(RP$output[[2]]$mrk[,3:7858])\n",
    "x<-as.matrix(n)-1\n",
    "data <- data.frame(tbcw=pheno,gid=1:2000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: foreach\n",
      "Loading required package: iterators\n",
      "Loading required package: parallel\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[1] '4.6'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "##load packages\n",
    "library(doParallel)\n",
    "library(foreach)\n",
    "cl<-makeCluster(8) \n",
    "repeats <- 10\n",
    "n.fold <- 5 \n",
    "acc<-list()\n",
    "library(rrBLUP)\n",
    "packageVersion(\"rrBLUP\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "#function for calculating the accuracy of GP with varying size of reference population\n",
    "cal_acc<-function(x,n.sample){\n",
    "id<-sample(1:dim(x)[1],n.sample)\n",
    "#relationship matrix (Endelman at al. 2011)\n",
    "A <- A.mat(x[id,], n.core=8)\n",
    "row.names(A)=1:n.sample;colnames(A)=1:n.sample\n",
    "data <- data.frame(tbcw=pheno[id],gid=1:n.sample)\n",
    "registerDoParallel(cl)\n",
    "GBLUP<-foreach(j=1:repeats,.combine = \"rbind\") %do% {\n",
    "        set.seed(100+3*j+1)\n",
    "        id <- sample(1:n.sample %% n.fold) + 1 \n",
    "        foreach(i=1:n.fold,.packages=\"rrBLUP\") %do% {\n",
    "         bcw_test <- data\n",
    "         bcw_test$tbcw[id == i] <- NA\n",
    "         res <- kin.blup(bcw_test, K=A, geno=\"gid\", pheno=\"tbcw\")\n",
    "         cor(data$tbcw[id==i],res$pred[id==i])\n",
    "    }\n",
    "   }\n",
    "stopImplicitCluster()\n",
    "acc<-unlist(GBLUP)\n",
    "return(acc)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "#implemente the caculation \n",
    "n.sample<-c(240,500,750,1000,1250,1500,1750,2000)\n",
    "all_acc<-matrix(NA,nrow = length(n.sample),ncol = repeats*n.fold)\n",
    "for (i in 1:length(n.sample)){\n",
    "    m<-n.sample[i]\n",
    "    all_acc[i,]<-cal_acc(x,m)}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {},
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAtAAAAHgCAMAAAC7G6qeAAAAD1BMVEUAAAAzMzNNTU3r6+v/\n//+EK80uAAAACXBIWXMAABJ0AAASdAHeZh94AAATRElEQVR4nO2djXarOLMFMZP3f+Z7T76E\nP2EZSd2bTlN7rck4tqhsoMLBwnamL0ISZbq7ACGWQWiSKghNUgWhSaogNEkVhCapgtAkVRCa\npMqI0P+1pXW8mEe9QLhWHkK746in5CG0O456Sh5Cu+Oop+QhtDuOekoeQrvjqKfkIbQ7jnpK\nHkK746in5CG0O456Sh5Cu+Oop+QhtDuOekqejdCEhAtHaCcc9ZQ8hHbHUU/JQ2h3HPWUPIR2\nx1FPycsg9Ov1MuVhTCDcA4V+vYyNxphAuOcJ/XpZG40xgXAIPR6MCYRD6PFgTCDc84TmHDoW\nD6GHwyxHJB5CjyfXLhHjktVDaHcc9ZQ8hHbHUU/JQ2h3HPWUvDuEtn4Sl2yXiHHJ6t0gtPk0\nW7JdIsYlq6cX2v5CSLJdIsYlq4fQ7jjqKXkI7Y6jnpKnF5pz6Fi4ZPVuEJpZjlC4ZPXuEPru\ndRbjqKfkIbQ7jnpKHkK746in5CG0O456Sh5Cu+Oop+QhtDuOekoeQrvjqKfkIbQ7jnpKHkK7\n46in5CG0O456Sh5Cu+Oop+QhtDvuYfWMX6fTykNod1x0Y4LXQ+hoOOsXFyJ0LQjtjTN/+TdC\n15JC6Mgvr7Z/gw5C14LQZRB6JAg9HoQeCUIjdEs4h5byELoMsxwjQejxxBY6ujHB6yH0cBB6\nJAg9HoQeSfB6CD0chB4JQo8HoUcSvB5CD+dhQhvzEHo8CB2Ih9DjCS50LmPEOIQeT/B6CF0L\nQpcJXg+ha0HoMsHrIXQtCF0meD2ErgWhywSvh9C1IHSZ4PUQuhaELhO8HkLXgtBlgtdD6FoQ\nukzweghdC0KXCV7vdqFfI/Guh9Blgte7X+gBU85+mOnvB0KXCV4vn9CWuPWxj0LP83z4frnZ\nuEWCGxO8HkLXcOtjn4Sef/77/XZrd+MWCW5M8HoIXcOtj30Qel6+fO1v/UvjFgluTPB6CF3D\nrY9dEXrxeO8zQleD0DUDQwg9ryfU3881+xuZ5O6f/yHG9azXtpk3JLQtr7Zok9CHE+rGX/Hg\nh8Dg9ThC13DrY21H6M03X81C375L6kHoDwv0C4jQd/AQ+sMCxlcKA8xyIHRLEDqi0Pt56LFz\n6Nt3ST0I/WGBHEL/TmzM228Q+kJuN8a6Xr+Agc6hq/HehFre/UIP7Iq/KLTlr+/6oxD6Nwit\nrYfQzjyE1tZDaGceQmvrcQ7tzEsn9Mgh8ISH0MUaNI4X826vF8iYM16gegh9Aw6h/eoh9A04\nhParh9A34BDarx5C34BDaL96f1Zo4+fp1vWMeYGMORXaeGc8UuiBcggdfOuZ/n6szRG6v54x\n72FCm/LW5gg9UM/2H2GE7uetzRE6TD3j349HPQNZNyxCh6mH0P28dcMidJh6CN3PWzcsQoep\nxzl0P29tjtBh6iF0P29tjtBh6iF0P29tjtBh6iF0P29tjtAD9WyfdSF0P29tjtAD9RB6YOuZ\n8tbmCD1QD6EHtp4pb22O0GHqIXQ/b22O0GHqIXQ/b22O0GHqIXQ/b22O0GHqBf8YA+utZ8pb\ntwRCh6kX/INmpDiEHt/Ft9eTHlER+iS+Hf97mtDW9aQ8hL60ROyzQIR2wyE0Qt/LQ+hLSyD0\nQD0pD6EvLRHbmOD1pDyEvrREbGOC15PyEPrSErGNCV5PykPoS0vENiZ4PSkPoS8tEduY4PWk\nPOt6jXKtGxahB+rFnoSR8hD60hKxhZbygtez/gs1CI3Qt/IQ+tISCO2FQ2iEvpcXvB5CX1oC\nob1wCI3Q9/KC10PoS0sgtBcOoTMIrZw3Rmgl76lC2+KsV1eJQ+iTNHZE6JEgdHX4EoTux1mv\nrhKH0Cdp7IjQI0Ho6vAlCN2Ps15dJQ6hT9LYEaFHgtDV4UsQuh9nvbpKHEKfpLEjQo8EoavD\nlyB0P856dZU4hD5JY0eEHglCV4cviS208bVqhHbjIfR4pEd8hK4nhdCtTtzOGxFa+cMEOHNe\nlHCEvpbbj9A3HwLFOE45ri2B0F48hB4PQgfiJRU6+CvoEdqNl1Xo2MYEr1cNQleHL0HoOPWq\nQejq8CUIHadeNQhdHb5kL/Q0tQhecoMbE7xeNQhdHb5kJ/B0vKOekhvcmOD1qkHo6vAlW3+n\n4p56Sm5wY4JPwlSD0NXhCI3QkXHWQl+TuuQi9Ei9ahC6OvyhQseuVw1CV4cv2Qm95tJOLrnB\njQlerxqErg5fUh6hj/e+T8kNbkzwetUgdHX4EoSOU68ahK4OX1Koyzz0bfWqQejq8CUI7VUv\nlzFinN2Twq/rzwkRuppcxohx1pe+OYcertczdV0LQleHLymeFCK0Sb2+izGVIHR1+BJmOVzq\n9V5efB+Erg5fgtAu9RBay1t3FEK71ENoLW/dUcWTwuLO9ym5CL38KM6hlbx1Rx3esXJy3/uU\nXIRefxazHELeuqMO8jbMQiN0PbmMEeN4k+y1JZT1chkjxiH0tSWU9XIZI8Yh9LUllPVyGSPG\nIfS1JZT1chkjxlnPQ19LyUXoNbmMEeNMhG4+WJdchF6Tyxgxzkzoqbi3kpKL0GtyGSPGIfS1\nJZT1chkjxvGu72tLKOvlMkaMs7tSeHrvm5RchF6Tyxgxjmm7a0soPzkplzFiXBihlcYYv63a\nGJfMGDHO8l3f11+dVHIRek0uY8Q4O6Eb3lKI0NXkMkaM40nheBA6EA6hx4PQgXAIPR6EDoQL\ncw6N0EtyGSPGhZnlQOgluYwR48LMQyP0klzGiHEIPR6EDoRD6PEgdCAcQo8HoQPhEHo8CB0I\nh9DjQehAOOsLKwg9nlzGiHF2Ql//IDCErieXMWKc5RGaCys2yWWMGGd8Dp31PYVKXDJjxDhb\nob//FhZCDyaXMWKc7SnHr9UIPZJcxohxHk8KwwttvA0ROhDOY9oOoceSyxgx7pEXVhA6EC+s\n0NN0cidCdySXMWKc41uw5nneDth+U3IRek0uY8Q4P6HnvcQzQl9NLmPEODeh5+XLz7cIfTW5\njBHj3N4kO2++InRTchkjxrm9SXYv9Lzc/P6soxMrB4TuX9QmtxcgFvkwobET+nA+zRG6mlyH\nQDHObR56Pn5F6KvJZYwYZy30/knh79efIPSl5DJGjLO8sLL7kxTFUZkj9NXkMkaMM3213f66\nSpvQyo/TRehAvNhCn1wp3Ex0VISuJrgxweshdHV4RejDIbqSxo7BjQleD6Grw98J/f3ipP4/\n61ZNcGOC10Po6vC3QreksWNwY4LXQ+jq8PdHaIQ2Si5jxDjDt2D9GaF7JkaqPFNaNmPEOI+3\nYAUXum+qrwa0hP2XzRgxzvqUI/6Twt7J6wrRkPUvuYwR4yzf9T3yx+urQehAuGT1OEKfEA1Z\n/5LLGDHuedN2CB2LF1vo7nd9V8OTwkC4ZPXeuvtnzqGZtovFCyv0nzlCc2ElFC+q0NV7Ebol\nuYwR4xB6PMHrIXR1OEIXCV4PoavD36nb8pwQoavJZYwY98DXciB0KF5UoZvS2DG4McHrIXR1\n+Fuhxz5Ot5rgxgSvh9DV4e+ELj9OF6E7k8sYMQ6hxxO8HkJXhyN0keD1ELo6/J3Qx4/TReju\n5DJGjDOc5bg+DY3Q1eQyRoxj2i4cjnpKHkK746in5L0X+rnz0BgTCPfEWQ7rF/hjTCDcA4U2\nfwsWxgTCPU9o+zfJYkwg3PPmoRE6Fi+s0H9lHhqhY/HiCt2Qxo6cQwfCJav3Sej459DMcsTi\nIfR4cu0SMS5Zvbq6f+ODZpLtEjEuWb2a0NND/2gQxgTCmV76vvw8sbEjQgfCJav3TuiWV3Ig\ntJRHverwU6Gnpk+2Q2gpj3rV4W+P0JdPoBFay6NedfgboducbuyI0IFwyepVhG54XtjYEaED\n4ZLVqwp9+a8VNnZE6EC4ZPU+CM0ROh6PetXhjeoi9O086lWHI7Q3jnpKHkK746in5CG0O456\nSh5Cu+Oop+QhtDuOekoeQrvjqKfkIbQ7jnpKHkK746in5CG0O456Sh5Cu+Oop+QhtDuOekoe\nQrvjqKfk2QjdmJfuR5HHhiO0E456Sh5Cu+Oop+QhtDuOekoeQrvjqKfkIXQR6w8zTWaMGIfQ\nozH/uOlkxohxCD0Y+z8IkMwYMQ6hB4PQsXAIPRiEjoVD6NFwDh0Kh9DDYZYjEg6hw+Gop+Qh\ntDuOekoeQrvjqKfkIbQ7jnpKHkK746in5CG0O456Sh5Cu+Oop+QhtDuOekoeQrvjqKfkIbQ7\njnpKHkK746in5CG0O456Sh5Cu+Oop+QhtDuOekoeQrvjqKfkIbQ7jnpKHkK746in5CG0O456\nSt4tQrNLAuGS1UNodxz1lDyEdsdRT8lDaHcc9ZQ8hHbHUU/JQ2h3HPWUPIR2x1FPyUNodxz1\nlDyEdsdRT8lDaHcc9ZQ8hHbHUU/JQ2h3HPWUPIR2x1FPyUNodxz1lDyEdsdRT8lDaHcc9ZQ8\nhHbHUU/JQ2h3HPWUPIR2x1FPyUNodxz1lDyEdsdRT8lDaHcc9ZQ8hHbHUU/JQ2h3HPWUPIR2\nx1FPyUNodxz1lDyEdsdRT8lDaHcc9ZQ8hHbHUU/JQ2h3HPWUPIR2x1FPyUNodxz1lDyEdsdR\nT8lrEHqe5+3tefOQb0c1j3qBcH5Czz//Lbc3Rvt2VPOoFwjnJvS8fDl+g9BaHvWqw5uE3kqM\n0DfxqFcd3i30z83Xv9QXJeSGdAr9Hd9fOjWPeoFwsiP01meEVvKoVx3eKfTOZ4RW8qhXHd4k\n9BufEVrJo151+FWhi3nordW+HdU86gXC+V8p/N9Flf2lQt+Oah71AuF4LUc4HPWUPIR2x1FP\nyUNodxz1lDyEdsdRT8lDaHcc9ZQ8hHbHUU/JQ2h3HPWUPIR2x1FPyUNodxz1lDyEdsdRT8lD\naHcc9ZQ8hHbHUU/JQ2h3HPWUPIR2x1FPyUNodxz1lDyEdsdRT8lDaHcc9ZQ8hHbHUU/JQ2h3\nHPWUPIR2x1FPyUNodxz1lDyEdsdRT8lDaHcc9ZQ8hHbHUU/JQ2h3HPWUPIR2x1FPyUNodxz1\nlDyEdsdRT8lDaHcc9ZQ8hHbHUU/JQ2h3HPWUPIR2x1FPyUNodxz1lDyEdsdRT8lDaHcc9ZQ8\nhHbHUU/JQ2h3HPWUPIR2x1FPyUNodxz1lDyEdsdRT8lDaHcc9ZQ8hHbHUU/JQ2h3HPWUPIR2\nx1FPybtD6Nfr1bbAx5UIjaOekneD0K+XtdG5dokYl6yeXujXy9zoXLtEjEtWD6HdcdRT8hDa\nHUc9JU8vNOfQsXDJ6t0gNLMcoXDJ6t0h9N3rLMZRT8lDaHcc9ZQ8hHbHUU/JQ2h3HPWUPIR2\nx1FPyUNodxz1lDyEdsdRT8lDaHcc9ZQ8hHbHUU/JsxGakHDhCO2Eo56Sh9DuOOopeQjtjqOe\nkofQ7jjqKXkI7Y6jnpJnI3RjXrF51AuE6+chtBOOevfwENoJR717eAjthKPePTyuFJJUQWiS\nKghNUgWhSaogNEkVhCap4iz0/P9Zbi93WQAXzABvnre4A7cDt2Lf3ejAzVvKSMdNPYtNuG43\ni004VzZaE89X6H9F1lrrXUPAPWaAtxY64bbTlhUtqvVA5+32Mui4q/emZzvOahPu13Jk6/kL\nvaz0fLinG/h1tspjyILbTXq3NzoOWodGox0PQtvUKxt14T65HEXo75RC9yuzPW59fx3j/S5a\ncrtRJ9U6oQdjhjseTwzM6lltwvVQPLD1/IXeNxsW8ICxEbrk9qMchLbpeDQwWL13rWIKffi3\nY/gcwUHoPbcf5Sn0UMfjET9Yvb8i9HzyJZDQ7/ZyP8teaKOOXkKb1gsv9E+d+Xd6B6Gbq82H\n+3IKPW//F1fo487YHqZHmHNxYwB2xu1nldU6obtdbNDx3aqO1TPahLvf3pF6gmm74pdv/AB9\nts4DsDNuP+x8dwz/GzzccT7cMqlntAn3S45sPWeh1+tI604ZuFK4pRlcKdx2Krg95T7eaMct\nm3C044b3tnA7zmgT7i84jtTjtRwkVRCapApCk1RBaJIqCE1SBaFJqiA0SRWEJqmC0NfzeVtN\n0/mYd/ePpqCyO9kCHzP90/HSdpq+zjfou/tHs/01YUf+L2yHT/neQteOsO8GuW1khC7CdviU\nafP10tDr9w8HoYuwHT5lFfr7v+987U+Lf25Py33T13LP5v6f776O301byAY8Lf8w/AK+fnjL\nl+XB31bH4Y/bv49b4fasVn1txP5at92q1rLE/9TbD1z9XR+bTqEbLafND5i2zOMCmzOj7fDH\n7eCnrW9XptXIr9Lf/SH8cNdh9HR87N3/T5coBh2F3v/67H7JnpKnrW9nNm4sZxfT1rtToaft\nicDvd/vH9tKegD8KvT2UI/Tj1rc9J6bsN1v9CF0ssn+sEPoI/iz0xBF6k6etb3sOphxcLQec\nDHIVmlOOXZ62vu2ZCn1+ZdlKuJdn2j+wk23/2Jmbx8VPH98+Q5yWg/Rp10flaevbnmmZEpum\nn1PcRajfIYdpu/UsuJy22z+2uetk2u7A2f+A7dLTtINvgI+buHvY6v6pFPuGnfU5bKO4QeiO\nsI3iBqE7wjYKm+L093Hnwz1hE5FUQWiSKghNUgWhSaogNEkVhCapgtAkVRCapMr/AbXbFnEY\nrWwhAAAAAElFTkSuQmCC",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Boxplot\n",
    "library(ggplot2)\n",
    "options(repr.plot.width=6, repr.plot.height=4)\n",
    "plot<-data.frame(correlation=as.vector(t(all_acc)),\n",
    "                 no_of_sample=as.factor(rep(n.sample,each = n.fold*repeats)))\n",
    "p <- ggplot(plot, aes(x=no_of_sample, y=correlation))+\n",
    "     geom_boxplot()+\n",
    "     labs(x = \"Size of reference population\",y=\"Accuracy of GP\")+\n",
    "     theme(legend.position=\"none\",\n",
    "           axis.text.y=element_text(size=10),\n",
    "           axis.text.x=element_text(size=10),\n",
    "           text=element_text(size=10, \n",
    "           family=\"serif\"))\n",
    "p"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {},
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Save\n",
    "png(\"2.1 Size of reference population.png\", width=6, height=4, units=\"in\", res=300)\n",
    "p\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th scope=col>no_of_sample</th><th scope=col>acc_mean</th><th scope=col>acc_se</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><td> 240       </td><td>0.4388824  </td><td>0.030182497</td></tr>\n",
       "\t<tr><td> 500       </td><td>0.5348484  </td><td>0.022595243</td></tr>\n",
       "\t<tr><td> 750       </td><td>0.4809595  </td><td>0.021344959</td></tr>\n",
       "\t<tr><td>1000       </td><td>0.5497540  </td><td>0.012919789</td></tr>\n",
       "\t<tr><td>1250       </td><td>0.5697551  </td><td>0.011575666</td></tr>\n",
       "\t<tr><td>1500       </td><td>0.5655269  </td><td>0.009653200</td></tr>\n",
       "\t<tr><td>1750       </td><td>0.5927709  </td><td>0.008615323</td></tr>\n",
       "\t<tr><td>2000       </td><td>0.5873600  </td><td>0.008948373</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|lll}\n",
       " no\\_of\\_sample & acc\\_mean & acc\\_se\\\\\n",
       "\\hline\n",
       "\t  240        & 0.4388824   & 0.030182497\\\\\n",
       "\t  500        & 0.5348484   & 0.022595243\\\\\n",
       "\t  750        & 0.4809595   & 0.021344959\\\\\n",
       "\t 1000        & 0.5497540   & 0.012919789\\\\\n",
       "\t 1250        & 0.5697551   & 0.011575666\\\\\n",
       "\t 1500        & 0.5655269   & 0.009653200\\\\\n",
       "\t 1750        & 0.5927709   & 0.008615323\\\\\n",
       "\t 2000        & 0.5873600   & 0.008948373\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "no_of_sample | acc_mean | acc_se | \n",
       "|---|---|---|---|---|---|---|---|\n",
       "|  240        | 0.4388824   | 0.030182497 | \n",
       "|  500        | 0.5348484   | 0.022595243 | \n",
       "|  750        | 0.4809595   | 0.021344959 | \n",
       "| 1000        | 0.5497540   | 0.012919789 | \n",
       "| 1250        | 0.5697551   | 0.011575666 | \n",
       "| 1500        | 0.5655269   | 0.009653200 | \n",
       "| 1750        | 0.5927709   | 0.008615323 | \n",
       "| 2000        | 0.5873600   | 0.008948373 | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "  no_of_sample acc_mean  acc_se     \n",
       "1  240         0.4388824 0.030182497\n",
       "2  500         0.5348484 0.022595243\n",
       "3  750         0.4809595 0.021344959\n",
       "4 1000         0.5497540 0.012919789\n",
       "5 1250         0.5697551 0.011575666\n",
       "6 1500         0.5655269 0.009653200\n",
       "7 1750         0.5927709 0.008615323\n",
       "8 2000         0.5873600 0.008948373"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#caculate the mean and se of accuracy of GP\n",
    "cal_se<-function(acc){sd(acc)/sqrt(repeats-1)}\n",
    "(sum<-data.frame(no_of_sample=n.sample,\n",
    "                 acc_mean = apply(all_acc, 1, mean),\n",
    "                 acc_se = apply(all_acc, 1, cal_se)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save to xlsx file\n",
    "library(\"xlsx\")\n",
    "write.xlsx(sum, \"2_1.xlsx\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
